pkg/trojan/message.go

```diff
@@ -0,0 +1,283 @@
+// and its data set to the serialization of the trojan chunk fields which correctly hash into the matching address
+func (m *Message) toChunk(targets Targets, span []byte) (swarm.Chunk, error) {
+	// start out with random nonce
+	nonce := make([]byte, 32)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, err
+	}
+	nonceInt := new(big.Int).SetBytes(nonce)
+	targetsLen := len(targets[0])
+
+	// serialize message
+	b, err := m.MarshalBinary() // TODO: this should be encrypted
+	if err != nil {
+		return nil, err
+	}
+
+	// hash chunk fields with different nonces until an acceptable one is found
+	// TODO: prevent infinite loop
+	for {
+		s := append(append(span, nonce...), b...) // serialize chunk fields
+		hash1, err := hashBytes(s)
+		if err != nil {
+			return nil, err
+		}
+
+		// take as much of the hash as the targets are long
+		if contains(targets, hash1[:targetsLen]) {
+			// if nonce found, stop loop and return chunk
+			return swarm.NewChunk(swarm.NewAddress(hash1), s), nil
+		}
+		// else, add 1 to nonce and try again
+		nonceInt.Add(nonceInt, big.NewInt(1))
+		// loop around in case of overflow
+		if nonceInt.BitLen() > 256 {
+			nonceInt = big.NewInt(0)
+		}
+		nonce = padBytes(nonceInt.Bytes()) // pad in case Bytes call is not 32 bytes long
+	}
+}
+
+// hashBytes hashes the given serialization of chunk fields with the hashing func
+func hashBytes(s []byte) ([]byte, error) {
+	if hasher == nil {
+		return nil, ErrInvalidHasher
+	}
+	hasher.Reset()
+	span := binary.LittleEndian.Uint64(s[:8])
+	err := hasher.SetSpan(int64(span))
+	if err != nil {
+		return nil, err
+	}
+	if _, err := hasher.Write(s[8:]); err != nil {
+		return nil, err
+	}
+	return hasher.Sum(nil), nil
+}
+
+// contains returns whether the given collection contains the given elem
+func contains(col Targets, elem []byte) bool {
+	for i := range col {
+		if bytes.Equal(elem, col[i]) {
+			return true
+		}
+	}
+	return false
+}
+
+// padBytes adds 0s to the given byte slice as left padding,
+// returning this as a new byte slice with a length of exactly 32
+// given param is assumed to be at most 32 bytes long
+func padBytes(b []byte) []byte {
+	l := len(b)
+	if l == 32 {
+		return b
+	}
+	bb := make([]byte, 32)
+	copy(bb[32-l:], b)
+	return bb
+}
+
+// MarshalBinary serializes a message struct
+func (m *Message) MarshalBinary() (data []byte, err error) {
+	data = append(m.length[:], m.Topic[:]...)
+	data = append(data, m.Payload...)
+	data = append(data, m.padding...)
+	return data, nil
+}
+
+// UnmarshalBinary deserializes a message struct
+func (m *Message) UnmarshalBinary(data []byte) (err error) {
+	copy(m.length[:], data[:2])  // first 2 bytes are length
+	copy(m.Topic[:], data[2:34]) // following 32 bytes are topic
+
+	// rest of the bytes are payload and padding
+	length := binary.BigEndian.Uint16(m.length[:])
+	payloadEnd := 34 + length
+	m.Payload = data[34:payloadEnd]
+	m.padding = data[payloadEnd:]
+	return nil
```

Comment:
It's a bit puzzling that unmarshaling cannot have an error condition. Please re-evaluate.

Reply:
it can `panic`. not sure we should add panic recovery code inside of an unmarshaling

Reply:
well it returns error in the signature for a reason ;)

Reply:
?

Reply:
Added validations, but github is not resolving this.

Reply:
do we want things to panic here?

Reply:
it's arguable. if none of the error `if` conditions are `true`, and we get an out of bounds error from line `274` onwards, then something has gone _very_ wrong.

`panic` might be preferable to silencing this problem. but i think a case could be made for both alternatives

Reply:
I added unmarshall Error

Reply:
@mortelli panicking then means that theoretically if an exploit can be found you can crash the node by passing bogus data to it. I would prefer a rash of errors in the logs instead.